cmd/snap/cmd_debug_timings.go

```diff
@@ -92,29 +93,70 @@ func printTaskTiming(w io.Writer, t *Timing, verbose, doing bool) {
 	printTiming(w, verbose, t.Level+1, "", "", doingTimeStr, undoingTimeStr, t.Label, t.Summary)
 }
 
-func (x *cmdChangeTimings) printChangeTimings(w io.Writer, timing *timingsData) error {
-	chgid := timing.ChangeID
-	chg, err := x.client.Change(chgid)
-	if err != nil {
-		return err
+// sortTimingsTasks sorts tasks from changeTimings by lane and ready time with special treatment of lane 0 tasks:
+// - tasks from lanes >0 are grouped by lanes and sorted by ready time.
+// - tasks from lane 0 are sorted by ready time and inserted before and after other lanes based on the min/max
+//   ready times of non-zero lanes.
+// - tasks from lane 0 with ready time between non-zero lane tasks are not really expected in our system and will
+//   appear after non-zero lane tasks.
+func sortTimingsTasks(timings map[string]changeTimings) []string {
+	tasks := make([]string, 0, len(timings))
+
+	var minReadyTime, maxReadyTime time.Time
+	for taskID, taskData := range timings {
+		if taskData.Lane > 0 {
+			if minReadyTime.IsZero() {
+				minReadyTime = taskData.ReadyTime
+				maxReadyTime = taskData.ReadyTime
+			}
+			if taskData.ReadyTime.Before(minReadyTime) {
+				minReadyTime = taskData.ReadyTime
+			}
+			if taskData.ReadyTime.After(maxReadyTime) {
+				maxReadyTime = taskData.ReadyTime
+			}
+		}
+		tasks = append(tasks, taskID)
 	}
 
-	for _, t := range chg.Tasks {
-		doingTime := formatDuration(timing.ChangeTimings[t.ID].DoingTime)
-		if timing.ChangeTimings[t.ID].DoingTime == 0 {
+	sort.Slice(tasks, func(i, j int) bool {
+		t1 := timings[tasks[i]]
+		t2 := timings[tasks[j]]
+		if t1.Lane != t2.Lane {
+			if t1.Lane == 0 {
+				return t1.ReadyTime.Before(minReadyTime)
+			}
+			if t2.Lane == 0 {
+				return t2.ReadyTime.After(maxReadyTime)
```

Comment:
I'm not sure I see where/how:
> // - tasks from lane 0 with ready time between non-zero lane tasks are not really expected in our system and will
> //   appear after non-zero lane tasks.

is implemented. Thinking quickly I think we would want t2.ReadyTime >= minReadyTime here to implement that. But that's my quick thought:  it means anything in lane 0 after the start of other lanes goes after anything in the != 0 lanes.

This also means we don't need maxReadyTime.

Reply:
Thanks, that's a good observation, I tweaked sorting function and updated respective test scenario (it would fail without the change).